### PR TITLE
feat: migrate table queries to typescript

### PR DESCRIPTION
**Pull request by lohart13**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

This PR migrates the following methods to typescript.
- `listSchemasQuery`
- `showAllTables` and `showTablesQuery`
- `dropTable` and `dropTableQuery`
- `dropAllTables`

Also, `showTablesQuery` has been renamed to `listTablesQuery` to keep the naming consistent.

### CODE REVIEW

**Comment by lohart13** [comment]
One question I have is should `showTablesQuery` be called `listTablesQuery` to match `listDatabasesQuery` and `listSchemasQuery`

**Comment by WikiRik** [comment]
> One question I have is should `showTablesQuery` be called `listTablesQuery` to match `listDatabasesQuery` and `listSchemasQuery`

That would be more consistent, go for it!

**Comment by lohart13** [comment]
> > One question I have is should `showTablesQuery` be called `listTablesQuery` to match `listDatabasesQuery` and `listSchemasQuery`
> 
> That would be more consistent, go for it!

Updated now

**Review by WikiRik — COMMENTED** [review]
Great work! Few remarks but they should be minor

File: packages/core/src/dialects/mssql/query-generator-typescript.ts [hunk 1]
```diff
@@ -15,6 +19,23 @@ const REMOVE_INDEX_QUERY_SUPPORTED_OPTIONS = new Set<keyof RemoveIndexQueryOptio
  * Temporary class to ease the TypeScript migration
  */
 export class MsSqlQueryGeneratorTypeScript extends AbstractQueryGenerator {
+  protected _getTechnicalSchemaNames() {
+    return ['INFORMATION_SCHEMA', 'sys'];
+  }
+
+  listSchemasQuery(options?: ListSchemasQueryOptions) {
+    const schemasToSkip = ['dbo', 'guest', ...this._getTechnicalSchemaNames()];
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/mssql/query-generator-typescript.ts:27]: No, it's not part of the default schemas in mssql.

File: packages/core/src/dialects/mariadb/query-generator-typescript.ts [hunk 2]
```diff
@@ -27,10 +31,40 @@ export class MariaDbQueryGeneratorTypeScript extends AbstractQueryGenerator {
     this.whereSqlBuilder.setOperatorKeyword(Op.notRegexp, 'NOT REGEXP');
   }
 
+  protected _getTechnicalSchemaNames() {
+    return ['MYSQL', 'INFORMATION_SCHEMA', 'PERFORMANCE_SCHEMA', 'SYS', 'mysql', 'information_schema', 'performance_schema', 'sys'];
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/mariadb/query-generator-typescript.ts:35]: When querying a list of all schemas for mariadb `sys` was returned by the query therefore I have excluded it to match the other dialects

File: packages/core/src/dialects/sqlite/query-generator-typescript.ts [hunk 3]
```diff
@@ -35,6 +35,20 @@ export class SqliteQueryGeneratorTypeScript extends AbstractQueryGenerator {
...
+      rejectInvalidOptions(
+        'listTablesQuery',
+        this.dialect.name,
+        LIST_TABLES_QUERY_SUPPORTABLE_OPTIONS,
+        LIST_TABLES_QUERY_SUPPORTED_OPTIONS,
+        options,
+      );
+    }
+
+    return 'SELECT name AS `tableName` FROM sqlite_master WHERE type=\'table\' AND name NOT LIKE \'sqlite_%\'';
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/sqlite/query-generator-typescript.ts:49]: Good point, I've updated the queries to only exclude the specific schemas. I don't have a complete list for ibmi but have listed what I could find.

File: packages/core/src/dialects/mariadb/query-generator-typescript.ts [hunk 4]
```diff
@@ -27,10 +31,40 @@ export class MariaDbQueryGeneratorTypeScript extends AbstractQueryGenerator {
...
 
+  listTablesQuery(options?: ListTablesQueryOptions) {
+    return joinSQLFragments([
+      'SELECT TABLE_NAME AS `tableName`,',
+      'TABLE_SCHEMA AS `schema`',
+      `FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_TYPE = 'BASE TABLE'`,
+      options?.schema
+        ? `AND TABLE_SCHEMA = ${this.escape(options.schema)}`
+        : `AND TABLE_SCHEMA NOT IN (${this._getTechnicalSchemaNames().map(schema => this.escape(schema)).join(', ')})`,
+      'ORDER BY TABLE_SCHEMA, TABLE_NAME',
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/mariadb/query-generator-typescript.ts:64]: Some dialects order them, so to keep the results consistent across all dialects, I added the order by.

File: packages/core/src/dialects/db2/query-generator-typescript.ts [hunk 5]
```diff
@@ -13,6 +17,23 @@ const REMOVE_INDEX_QUERY_SUPPORTED_OPTIONS = new Set<keyof RemoveIndexQueryOptio
...
+  listSchemasQuery(options?: ListSchemasQueryOptions) {
+    const schemasToSkip = this._getTechnicalSchemaNames();
+
+    if (options && Array.isArray(options?.skip)) {
+      schemasToSkip.push(...options.skip);
+    }
+
+    return joinSQLFragments([
+      'SELECT SCHEMANAME AS "schema" FROM SYSCAT.SCHEMATA',
+      `WHERE SCHEMANAME NOT LIKE 'SYS%' AND SCHEMANAME NOT IN (${schemasToSkip.map(schema => this.escape(schema)).join(', ')})`,
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/src/dialects/db2/query-generator-typescript.ts:33]: I've updated the queries now.

File: packages/core/test/unit/query-generator/list-tables-query.test.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { buildInvalidOptionReceivedError } from '@sequelize/core/_non-semver-use-at-your-own-risk_/utils/check.js';
+import { expectsql, sequelize } from '../../support';
+
+describe('QueryGenerator#listTablesQuery', () => {
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/unit/query-generator/list-tables-query.test.ts:4]: Done

File: packages/core/test/integration/query-interface/show-all-tables.test.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          { tableName: 'my_test_table2', schema: 'dummy_db' },
+          { tableName: 'my_test_table1', schema: sequelize.dialect.getDefaultSchema() },
+        ]);
+      });
+    }
+  });
+
+  if (sequelize.dialect.supports.schemas) {
+    describe('with schemas', () => {
+      it('should show all tables selected schemas', async () => {
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/integration/query-interface/show-all-tables.test.ts]: Updated now.

File: packages/core/test/integration/query-interface/drop-table.test.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            references: {
+              table: 'levels',
+              key: 'id',
+            },
+          },
+        });
+
+        await queryInterface.dropTable('levels', { cascade: true });
+        const exists = await queryInterface.tableExists('levels');
+        expect(exists).to.be.false;
```
[the remark this answers is not in the document]
Comment (lohart13) [reply packages/core/test/integration/query-interface/drop-table.test.ts]: Updated test now.

**Comment by lohart13** [comment]
Tagging PR to https://github.com/sequelize/website/issues/501

**Pull request merged**